### PR TITLE
PMap replacements

### PR DESCRIPTION
**Pull request by orange-vertex**
Depends on PR https://github.com/SpaceGroupUCL/depthmapX/pull/139

### CODE REVIEW

File: salalib/shapemap.cpp [hunk 1]
```diff
@@ -717,7 +720,7 @@ bool ShapeMap::moveShape(int shaperef, const Line& line, bool undoing)
       m_attributes.setValue(rowid, conn_col, (float) connectivity );
       if (isAxialMap()) {
          leng_col = m_attributes.getOrInsertLockedColumnIndex("Line Length");
-         m_attributes.setValue(rowid, leng_col, (float) m_shapes[rowid].getLength() );
+         m_attributes.setValue(rowid, leng_col, (float) m_shapes.find(rowid)->second.getLength() );
```
Comment (blsemo) [inline salalib/shapemap.cpp]: key/index?

File: salalib/shapemap.cpp [hunk 2]
```diff
@@ -1165,7 +1087,7 @@ void ShapeMap::undo()
          //
          if (event.m_geometry.isLine()) {
             int leng_col = m_attributes.getOrInsertLockedColumnIndex("Line Length");
-            m_attributes.setValue(rowid, leng_col, (float) m_shapes[rowid].getLength() );
+            m_attributes.setValue(rowid, leng_col, (float) m_shapes.find(rowid)->second.getLength() );
```
Comment (blsemo) [inline salalib/shapemap.cpp]: key/index?

File: salalib/shapemap.cpp [hunk 3]
```diff
@@ -1812,14 +1735,14 @@ void ShapeMap::shapeInPolyList(const SalaShape& shape, pvecint& shapeindexlist)
...
+   auto shapeIter = m_shapes.end();
    // simplist: in shape centre
    if (shape.m_tags & ShapeRef::SHAPE_CENTRE) {
-      shapeindex = m_shapes.searchindex(shape.m_shape_ref);
+      shapeIter = m_shapes.find(shape.m_shape_ref);
    }
    // check not an open shape (cannot be inside)
    else if ((shape.m_tags & ShapeRef::SHAPE_OPEN) == 0) {
-      const SalaShape& poly = m_shapes.search(shape.m_shape_ref);
+      const SalaShape& poly = shapeIter->second;
```
Comment (blsemo) [inline salalib/shapemap.cpp]: Erm, wouldn't that dereference `m_shapes.end()`, which would throw if ever invoked?

File: salalib/shapemap.cpp [hunk 4]
```diff
@@ -1970,13 +1893,14 @@ int ShapeMap::getClosestOpenGeom(const Point2f& p) const
...
-   size_t index = paftl::npos;
+   auto shapeIter = m_shapes.end();
    double mindist = -1;
    for (size_t i = 0; i < m_pixel_shapes[pix.x][pix.y].size(); i++) {
       ShapeRef& ref = m_pixel_shapes[pix.x][pix.y][i];
       if (ref.m_tags & ShapeRef::SHAPE_OPEN) {
          double thisdist = -1.0;
-         const SalaShape& poly = m_shapes.search(ref.m_shape_ref);
+         auto tempShapeIter = m_shapes.find(ref.m_shape_ref);
+         const SalaShape& poly = shapeIter->second;
```
Comment (blsemo) [inline salalib/shapemap.cpp]: I think you wanted to use `tempShapeIter` here...

File: salalib/shapemap.cpp [hunk 5]
```diff
@@ -2135,7 +2059,7 @@ void ShapeMap::getShapeCuts(const Line& li_orig, pvector<ValuePair>& cuts)
             else {
                // this is a non-poly, so check just the shape_ref:
                if (tested.searchindex(IntPair(shaperef.m_shape_ref,-1)) == paftl::npos) {
-                  SalaShape& poly = m_shapes.search(shaperef.m_shape_ref);
+                  SalaShape& poly = m_shapes.find(shaperef.m_shape_ref)->second;
```
Comment (blsemo) [inline salalib/shapemap.cpp]: [] operator?

File: salalib/shapemap.cpp [hunk 6]
```diff
@@ -2228,8 +2152,9 @@ int ShapeMap::withinRadius(const Point2f& pt, double radius, std::vector<int>& b
             if (len == 0) {
                // this is a non-poly, so check just the shape_ref:
                if (tested.searchindex(IntPair(shaperef.m_shape_ref,-1)) == paftl::npos) {
-                  size_t shapeindex = m_shapes.searchindex(shaperef.m_shape_ref);
-                  SalaShape& poly = m_shapes[shapeindex];
+                  auto shapeIter = m_shapes.find(shaperef.m_shape_ref);
+                  size_t shapeindex = std::distance(m_shapes.begin(), shapeIter);
+                  SalaShape& poly = shapeIter->second;
                   if (poly.isPoint() && dist(pt,poly.getPoint()) < radius) {
                      depthmapX::addIfNotExists(bufferset, int(shapeindex));
```
Comment (blsemo) [inline salalib/shapemap.cpp]: Hard to tell from this review, but couldn't `bufferset` be a `std::set`?

File: salalib/shapemap.cpp [hunk 7]
```diff
@@ -2291,7 +2217,7 @@ int ShapeMap::connectIntersected(int rowid, bool linegraph)
       getShapeConnections( shaperef, m_connectors[rowid].m_connections, TOLERANCE_B*__max(m_region.height(),m_region.width()));
    m_attributes.setValue(rowid, conn_col, (float) connectivity );
    if (linegraph) {
-      m_attributes.setValue(rowid, leng_col, (float) m_shapes[rowid].getLength() );
+      m_attributes.setValue(rowid, leng_col, (float) m_shapes.find(rowid)->second.getLength() );
```
Comment (blsemo) [inline salalib/shapemap.cpp]: Aren't you trying to find the same shape again as you did in line 2203?

File: salalib/shapemap.cpp [hunk 8]
```diff
@@ -2763,10 +2695,10 @@ bool ShapeMap::output( ofstream& stream, char delimiter, bool updated_only )
       if (m_attributes.isVisible(i)) {
          stream << m_attributes.getRowKey(i);
          if ((m_map_type & LINEMAP) == 0) {
-            stream << delimiter << m_shapes[i].m_centroid.x << delimiter << m_shapes[i].m_centroid.y;
+            stream << delimiter << m_shapes.find(i)->second.m_centroid.x << delimiter << m_shapes.find(i)->second.m_centroid.y;
```
Comment (blsemo) [inline salalib/shapemap.cpp]: key/index?

File: salalib/shapemap.cpp [hunk 9]
```diff
@@ -2763,10 +2695,10 @@ bool ShapeMap::output( ofstream& stream, char delimiter, bool updated_only )
       if (m_attributes.isVisible(i)) {
          stream << m_attributes.getRowKey(i);
          if ((m_map_type & LINEMAP) == 0) {
-            stream << delimiter << m_shapes[i].m_centroid.x << delimiter << m_shapes[i].m_centroid.y;
+            stream << delimiter << m_shapes.find(i)->second.m_centroid.x << delimiter << m_shapes.find(i)->second.m_centroid.y;
          }
          else {
-            const Line& li = m_shapes[i].getLine();
+            const Line& li = m_shapes.find(i)->second.getLine();
```
Comment (blsemo) [inline salalib/shapemap.cpp]: key/index?

File: salalib/shapemap.cpp [hunk 10]
```diff
@@ -438,13 +438,13 @@ int ShapeMap::makePolyShape(const pqvector<Point2f>& points, bool open, bool tem
...
-      m_shapes.add(m_shape_ref,SalaShape(SalaShape::SHAPE_POLY));
+      m_shapes.insert(std::make_pair(m_shape_ref,SalaShape(SalaShape::SHAPE_POLY)));
    }
    else {
-      m_shapes.add(m_shape_ref,SalaShape(SalaShape::SHAPE_POLY | SalaShape::SHAPE_CLOSED));
+      m_shapes.insert(std::make_pair(m_shape_ref,SalaShape(SalaShape::SHAPE_POLY | SalaShape::SHAPE_CLOSED)));
    }
    for (i = 0; i < len; i++) {
-      m_shapes.tail().push_back(points[i]);
+      m_shapes.rbegin()->second.push_back(points[i]);
```
[the remark this answers is not in the document]
Comment (orange-vertex) [reply salalib/shapemap.cpp]: std::map doesn't have that function...

**Pull request merged**